signer/src/storage/postgres.rs

```diff
@@ -361,10 +396,74 @@ impl super::DbRead for PgStore {
 
     async fn get_pending_withdraw_requests(
         &self,
-        _chain_tip: &model::BitcoinBlockHash,
-        _context_window: usize,
+        chain_tip: &model::BitcoinBlockHash,
+        context_window: i32,
     ) -> Result<Vec<model::WithdrawRequest>, Self::Error> {
-        Ok(Vec::new()) // TODO(246): Write query + integration test
+        let stacks_chain_tip = self.get_stacks_chain_tip(chain_tip).await?;
+        sqlx::query_as!(
+            model::WithdrawRequest,
+            r#"
+            WITH RECURSIVE extended_context_window AS (
+                SELECT block_hash, parent_hash, confirms, created_at, 1 AS depth
+                FROM sbtc_signer.bitcoin_blocks
+                WHERE block_hash = $1
+                UNION ALL
+                SELECT
+                    parent.block_hash
+                  , parent.parent_hash
+                  , parent.confirms
+                  , parent.created_at
+                  , last.depth + 1
+                FROM sbtc_signer.bitcoin_blocks parent
+                JOIN extended_context_window last ON parent.block_hash = last.parent_hash
+                WHERE last.depth <= $3 + 1
+            ),
+            last_bitcoin_block AS (
+                SELECT
+                    block_hash
+                  , confirms
+                FROM extended_context_window
+                WHERE depth = $3 + 1
+            ),
+            StacksContext AS (
+                SELECT
+                    stacks_blocks.block_hash
+                  , stacks_blocks.block_height
+                  , stacks_blocks.parent_hash
+                  , stacks_blocks.created_at
+                FROM sbtc_signer.stacks_blocks stacks_blocks
+                WHERE stacks_blocks.block_hash = $2
+                UNION ALL
+                SELECT
+                    parent.block_hash
+                  , parent.block_height
+                  , parent.parent_hash
+                  , parent.created_at
+                FROM sbtc_signer.stacks_blocks parent
+                JOIN StacksContext last
+                        ON parent.block_hash = last.parent_hash
+                LEFT JOIN last_bitcoin_block block
+                        ON block.confirms @> ARRAY[parent.block_hash]
+                WHERE block.block_hash IS NULL
```

Comment:
Ohh I see. So here we fetch all stacks blocks that start at the latest stacks block that has been confirmed by a bitcoin block and end at the end of the context window.

I had thought that it would be simpler than that, where we fetch all stacks blocks starting at the chain tip (regardless of whether it was confirmed by a bitcoin block) and end at the end of the context window.

Reply:
Also, I think you may have mentioned this recently, but if we want stacks blocks from the latest stacks block confirmed on bitcoin, why not do something like the following, which explicitly joins on the stacks blocks confirmed on bitcoin starting from the bitcoin chain tip.
```sql
WITH RECURSIVE bitcoin_context_window AS (
    SELECT
        block_hash
      , parent_hash
      , UNNEST(confirms) AS stacks_hash
      , 1                AS depth
    FROM sbtc_signer.bitcoin_blocks AS bb
    WHERE block_hash = -- chain tip hash below
          '\xF13F8F587B670909CB8EDF864838D579608A2A8E531841C78595ADA1A3E60196'

    UNION ALL

    SELECT
        parent.block_hash
      , parent.parent_hash
      , UNNEST(parent.confirms) AS stacks_hash
      , last.depth + 1
    FROM sbtc_signer.bitcoin_blocks parent
    JOIN bitcoin_context_window AS last
         ON parent.block_hash = last.parent_hash
    WHERE last.depth <= 8 -- + 1
)
SELECT
    wr.request_id
  , wr.block_hash
  , wr.recipient
  , wr.amount
  , wr.max_fee
  , wr.sender_address
  , wr.created_at
FROM sbtc_signer.withdraw_requests wr
JOIN bitcoin_context_window sc ON wr.block_hash = sc.stacks_hash
```

Reply:
Thank you for suggesting an alternative query. I ran the `should_return_the_same_pending_withdraw_requests_as_in_memory_store` test with this query now, and this suggestion under-fetches withdraw requests.

The way I understand this is that it joins the bitcoin context window with the stacks blocks confirmed by the blocks in the context window. In that case we're missing all the stacks blocks that lies between stacks blocks confirmed by bitcoin, which should explain the under-fetching.

However, if you can think of a more clean/elegant formulation of the query that also joins the stacks blocks between the bitcoin blocks I'm happy to update it.

Reply:
> The way I understand this is that the suggested query joins the bitcoin context window with the stacks blocks confirmed by the blocks in the context window. In that case we're missing all the stacks blocks that lies between stacks blocks confirmed by bitcoin, which should explain the under-fetching.

Oh interesting, there could be blocks between the stacks blocks confirmed by bitcoin? Did not know that, so glad I brought it up.

Reply:
But I still have the question from above which blocks to fetch https://github.com/stacks-network/sbtc/pull/275#discussion_r1662925032.

Reply:
lol you answered it below. never mind.